Userland/Libraries/LibGL/SoftwareGLContext.cpp

```diff
@@ -208,175 +207,53 @@ void SoftwareGLContext::gl_end()
-
-    auto mvp = m_projection_matrix * m_model_view_matrix;
-
-    // Now let's transform each triangle and send that to the GPU
-    for (size_t i = 0; i < triangle_list.size(); i++) {
-        GLTriangle& triangle = triangle_list.at(i);
-
-        // First multiply the vertex by the MODELVIEW matrix and then the PROJECTION matrix
-        triangle.vertices[0].position = mvp * triangle.vertices[0].position;
-        triangle.vertices[1].position = mvp * triangle.vertices[1].position;
-        triangle.vertices[2].position = mvp * triangle.vertices[2].position;
-
-        // Apply texture transformation
-        // FIXME: implement multi-texturing: texcoords should be stored per texture unit
-        triangle.vertices[0].tex_coord = m_texture_matrix * triangle.vertices[0].tex_coord;
-        triangle.vertices[1].tex_coord = m_texture_matrix * triangle.vertices[1].tex_coord;
-        triangle.vertices[2].tex_coord = m_texture_matrix * triangle.vertices[2].tex_coord;
-
-        // At this point, we're in clip space
-        // Here's where we do the clipping. This is a really crude implementation of the
-        // https://learnopengl.com/Getting-started/Coordinate-Systems
-        // "Note that if only a part of a primitive e.g. a triangle is outside the clipping volume OpenGL
-        // will reconstruct the triangle as one or more triangles to fit inside the clipping range. "
-        //
-        // ALL VERTICES ARE DEFINED IN A CLOCKWISE ORDER
-
-        // Okay, let's do some face culling first
-
-        m_clipped_vertices.clear_with_capacity();
-        m_clipped_vertices.append(triangle.vertices[0]);
-        m_clipped_vertices.append(triangle.vertices[1]);
-        m_clipped_vertices.append(triangle.vertices[2]);
-        m_clipper.clip_triangle_against_frustum(m_clipped_vertices);
-
-        if (m_clipped_vertices.size() < 3)
-            continue;
-
-        for (auto& vec : m_clipped_vertices) {
-            // perspective divide
-            float w = vec.position.w();
-            vec.position.set_x(vec.position.x() / w);
-            vec.position.set_y(vec.position.y() / w);
-            vec.position.set_z(vec.position.z() / w);
-            vec.position.set_w(1 / w);
-
-            // to screen space
-            vec.position.set_x(scr_width / 2 + vec.position.x() * scr_width / 2);
-            vec.position.set_y(scr_height / 2 - vec.position.y() * scr_height / 2);
-        }
-
-        GLTriangle tri;
-        tri.vertices[0] = m_clipped_vertices[0];
-        for (size_t i = 1; i < m_clipped_vertices.size() - 1; i++) {
-
-            tri.vertices[1] = m_clipped_vertices[i];
-            tri.vertices[2] = m_clipped_vertices[i + 1];
-            processed_triangles.append(tri);
-        }
-    }
-
-    m_bound_texture_units.clear();
-    for (auto& texture_unit : m_texture_units) {
-        if (texture_unit.is_bound())
-            m_bound_texture_units.append(texture_unit);
+    Vector<size_t, 32> enabled_texture_units;
```

Comment:
Can we centralize the total number of texture units (32) in a `static constexpr` somewhere?

Reply:
Agreed!